### PR TITLE
Update mock storage

### PR DESCRIPTION
**Pull request by aljones15**
Addresses a bug found by Orie in the unit tests: https://github.com/digitalbazaar/edv-client/issues/17

1. Attributes are added to updated indexes if no updated attributes were found.
2. Ensure the indexed document is the latest version of the document.
3. Adds 2 tests from Orie that ensure the test framework works as expected.

```sh
EdvClient
    ✓ should only contain one indexed document after 3 updates (91ms)
    ✓ should have matching sequence numbers in indexes
    ✓ should create a new encrypted data vault
    ✓ should get an encrypted data vault config
    ✓ should create "primary" encrypted data vault
    ✓ should get "primary" encrypted data vault
    ✓ should ensure two new indexes
    ✓ should insert a document
    ✓ should get a document
    ✓ should fail to get a non-existent document
    ✓ should fail to insert a duplicate document
    ✓ should upsert a document
    ✓ should update an existing document
    ✓ should delete an existing document
    ✓ should fail to delete a non-existent document
    ✓ should insert a document with attributes
    ✓ should reject two documents with same unique attribute
    ✓ should find a document that has an attribute
    ✓ should find two documents with an attribute
    ✓ should find a document that equals an attribute value
    ✓ should find a document that equals the value of a URL attribute
    ✓ should find a document with a deep index on an array
    ✓ should find two documents with attribute values

  EdvDocument
    ✓ should read a document using EdvDocument


  24 passing (461ms)
```

### CODE REVIEW

File: tests/10-EdvClient.spec.js [hunk 1]
```diff
@@ -16,6 +16,106 @@ describe('EdvClient', () => {
...
+      content: {someKey: '111111', indexedKey: 'value1'}
+    };
+    let version1 = await client.insert({
+      doc,
+      invocationSigner,
+      keyResolver
+    });
+    version1 = await client.get({id: doc.id, invocationSigner});
+
+    // console.log("version: 1", version1);
```
Comment (OR13) [inline tests/10-EdvClient.spec.js]: nite: you may wish to strip my lazy console logs

File: tests/10-EdvClient.spec.js [hunk 2]
```diff
@@ -16,6 +16,106 @@ describe('EdvClient', () => {
...
+        content: {
+          ...version2.content,
+          someKey: '33333'
+        }
+      },
+      invocationSigner,
+      keyResolver
+    });
+
+    const docsFromGet = await getLatestDocsByIndexHas(
```
Comment (OR13) [inline tests/10-EdvClient.spec.js]: My original confusion was whether the result of this custom function should be equal to the direct client call:

```
const indexedDocs = await client.find({
        has: 'content.indexedKey,
        invocationSigner
      });
```

Comment (OR13) [reply tests/10-EdvClient.spec.js]: If its the case that the index should return 1 document, you should remove my hacky getLatestDocsByIndexHas function from this test.

File: tests/MockStorage.js [hunk 3]
```diff
@@ -180,8 +180,18 @@ export class MockStorage {
...
-
+    const oldDoc = edv.documents.get(doc.id);
     edv.documents.set(doc.id, doc);
+    if(oldDoc) {
+      // if an indexed value was not updated
+      // it's attributes is an empty array.
+      // so we need to use the last attributes.
+      for(const index in doc.indexed) {
+        if(doc.indexed[index].attributes.length === 0) {
+          doc.indexed[index].attributes = oldDoc.indexed[index].attributes;
```
Comment (dlongley) [inline tests/MockStorage.js]: I don't understand what's going on here. If the new doc doesn't have attributes we shouldn't be adding them in from an old document ... the new doc should be honored. This is an update where the attributes were removed from the new doc, right?

I would have expected this patch to go through the old doc's indexed attributes and remove it from all indexes... and then just run the rest of the code as normal, which would add the new doc to any indexed attributes (which may be none). *Perhaps* this change has that effect, but if so, it seems convoluted how it gets there.

Comment (aljones15) [reply tests/MockStorage.js]: 1. it's not a new doc it's an update to an existing doc.
2. the update does not change a property that is indexed.
3. because an indexed property is not changed no attributes are generated.

These were the cases I was considering in this patch:
Case 1:
1. User updates an existing document.
2. User's update changes an indexed property.
3. Update produces a new attributes array for the indexed field.
4. We check and do not overwrite the new attributes.
5. new attribute is indexed with latest document.

Case 2:
1. User updates an exiting document.
2. User's update does not change an indexed property.
3. The update passed to storage here will have an empty attributes array.
4. Because attributes is empty the indexed document is not updated (this is Orie's bug)
5. So we preserve the original attributes in our mock database resulting in an updated document.

I'm not aware of the ability to delete previous indexes however laying out the cases I do believe has revealed a minor new bug: if you change the attribute.value the key used to index changes which might result in the old unused key staying. While that does not result in test failure it does mean the test database preserves old keys. Although when I think about it not sure if mongo deletes the old indexed keys itself.

Basically this code is taking on the job of the database. I believe mongo is not destructive on update.

Comment (dlongley) [reply tests/MockStorage.js]: > it's not a new doc it's an update to an existing doc.
the update does not change a property that is indexed.
because an indexed property is not changed no attributes are generated.

An update replaces an existing document entirely. There is no patch functionality. A document that is sent as an update that does not include a previously indexed field should cause that field to be dropped.

So, in short, I don't think those cases exist at all. There's no concept of a patch/partial update in the API, there is only total replacement.

Comment (aljones15) [reply tests/MockStorage.js]: @dlongley I'm really sorry I totally forgot the behavior of this library even though I did use it before! I believe that orie's second test (`should have matching sequence numbers in indexes`) should be failing then because:

```
    client.ensureIndex({attribute: 'content.indexedKey'});
    const testId = await EdvClient.generateId();
    const doc = {
      id: testId,
      content: {someKey: 'someValue', indexedKey: 'value1'}
    };
    const version1 = await client.insert({
      doc,
      invocationSigner,
      keyResolver
    });
    version1.content = {someKey: 'aNewValue'};
    await client.update({doc: version1, invocationSigner, keyResolver});
```

then the second update to the same doc removes the `indexedKey`. 

`should only contain one indexed document after 3 updates`
keeps updates with the indexedKey in the doc.content so it should succeed.

`should have matching sequence numbers in indexes` however replaces the indexedKey with nothing so there should be no document in the old index hence it should be failing.

So basically if the indexed property is removed in an update, the indexes that store that document should also be removed. Or what you said in the first place which is just replace the index entirely.

Comment (dlongley) [reply tests/MockStorage.js]: @aljones15, no worries! Yeah, that second update removes `indexedKey` so the doc should get dropped from the index. Reading the rest sounds like you've got a handle on things now. Thanks!

Comment (aljones15) [reply tests/MockStorage.js]: @dlongley this now removes the indexed documents on update: https://github.com/digitalbazaar/edv-client/pull/18/commits/91e6ac4cf2d27970276c9e42eec616268dd7c13f

File: tests/10-EdvClient.spec.js (hunk 2 above)
Comment (aljones15) [reply tests/10-EdvClient.spec.js]: @OR13 this removes your hack function https://github.com/digitalbazaar/edv-client/pull/18/commits/91e6ac4cf2d27970276c9e42eec616268dd7c13f

However one of your tests should be failing.
When you update an EDV document you either update the whole thing or nothing.
So your smaller test where on update you do not have `indexedKey` in the content on update means the there is no longer an `indexedKey` for that document. So it should have been failing, but was failing for the wrong reason! Basically EDV documents do not support partial updates the way mongo documents normally do. Primary reason for this is probably because your document is stored encrypted in the database and mongo does not have operators that can work on encrypted data yet. Also it's just simple and easy to manage.

File: tests/10-EdvClient.spec.js (hunk 1 above)
Comment (aljones15) [reply tests/10-EdvClient.spec.js]: Nit should be resolved: https://github.com/digitalbazaar/edv-client/pull/18/commits/91e6ac4cf2d27970276c9e42eec616268dd7c13f

File: tests/10-EdvClient.spec.js [hunk 4]
```diff
@@ -16,6 +16,90 @@ describe('EdvClient', () => {
     await mock.server.shutdown();
   });
 
+  it('should find document by index after updates', async () => {
+
```
Comment (dlongley) [inline tests/10-EdvClient.spec.js]:
```suggestion
```

File: tests/10-EdvClient.spec.js (hunk 2 above)
Comment (OR13) [reply tests/10-EdvClient.spec.js]: Thanks, this make things much clearer, when performing an update, any indexes that need to be applied need to be maintained, manually.
